DF.__setitem__ creates extension column when given extension scalar

justinessert opened the pull request:
This PR is in response to [Issue 34832]( https://github.com/pandas-dev/pandas/issues/34832).
These changes follow the suggestions from reviewers on the PR to fix an issue where `df['a']=pd.Period('2020-01')` would create an object column instead of a `period[M]` column.

One potential issue that I see with these changes is that this code requires the `shape` parameter passed into `cast_scalar_to_array` to be an int, even though the functions docstring claims that it accepts a tuple.

I'm happy to work on resolving that issue, but first wanted to clarify that it is necessary. Based on my perspective, there would be no reason to call this function to create a multi-dimensional array, but I surely don't understand the full interconnectivity of this package.

- [x] closes #xxxx
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

justinessert committed:
Bugfix to make DF.__setitem__ create extension column instead of object column when given an extension scalar

pep8speaks commented:
Hello @justinessert! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`pandas/core/dtypes/cast.py`](https://github.com/pandas-dev/pandas/blob/0ec5911ff88b96b8b8ee906dd135774e618525e6/pandas/core/dtypes/cast.py):

> [Line 1512:1](https://github.com/pandas-dev/pandas/blob/0ec5911ff88b96b8b8ee906dd135774e618525e6/pandas/core/dtypes/cast.py#L1512): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace

justinessert committed:
removed bad whitespace

TomAugspurger commented on pandas/core/dtypes/cast.py:
```diff
@@ -1505,12 +1506,15 @@ def cast_scalar_to_array(shape, value, dtype: Optional[DtypeObj] = None) -> np.n
 
-    values = np.empty(shape, dtype=dtype)
-    values.fill(fill_value)
+    if type(dtype) in registry.dtypes:
```
```suggestion
    if is_extension_array_dtype(dtype)
```

TomAugspurger commented on pandas/core/dtypes/cast.py:
```diff
@@ -60,6 +60,7 @@
     ExtensionDtype,
     IntervalDtype,
     PeriodDtype,
+    registry
```
```suggestion
```

TomAugspurger commented on pandas/core/dtypes/cast.py:
```diff
@@ -1505,12 +1506,15 @@ def cast_scalar_to_array(shape, value, dtype: Optional[DtypeObj] = None) -> np.n
     """
     if dtype is None:
-        dtype, fill_value = infer_dtype_from_scalar(value)
+        dtype, fill_value = infer_dtype_from_scalar(value, pandas_dtype=True)
```
glancing through `infer_dtype_from_scalar`, it doesn't look like this will work for 3rd party extension types.

That's probably an OK limitation for now, but we'll need to address it some day. It's complicated by the fact that it might be ambiguous which extension type a scalar belongs to (we might have multiple extension arrays that could hold, e.g. integers).

justinessert replied:
I see, I can add a comment about this as a future TODO?
```suggestion
        # TODO: Update this function to add support for 3rd party extension types
        dtype, fill_value = infer_dtype_from_scalar(value, pandas_dtype=True)
```

TomAugspurger replied:
Yeah, can you open an issue and link to it in the todo? You can use one of the example extension arrays in `pandas/tests/extension` like `DecimalArray` to demonstrate.

justinessert replied:
Done, added the comment and created [Issue 34959](https://github.com/pandas-dev/pandas/issues/34959)

TomAugspurger commented on pandas/core/dtypes/cast.py:
```diff
@@ -1505,12 +1506,15 @@ def cast_scalar_to_array(shape, value, dtype: Optional[DtypeObj] = None) -> np.n
-    values = np.empty(shape, dtype=dtype)
-    values.fill(fill_value)
+    if type(dtype) in registry.dtypes:
+        values = dtype.construct_array_type()._from_sequence([value] * shape)
```
What's the type on shape here? A tuple or integer?

Extension arrays are only allowed to be 1D. We'll need to make sure that the caller is able to handle this. May need to restructure the logic a bit to make sure it's not surprised.

justinessert replied:
Yeah, this is the part that was most confusing to me.

Correct me if I'm wrong, but I believe that the motivation for this function accepting a tuple instead of an int is that similar data type columns can be grouped together in memory by creating them all within the same numpy array.

If that's correct, and if Extension arrays can only be 1D, then the same logic would not really apply to extension types, right? There would be no reason to call this function using `shape` as a tuple, since no grouping could happen anyway.

To be explicit, what I'm suggesting is perhaps only allowing `shape` to be an int when `value` is an extension type, otherwise we throw an error. We could also build in support if `shape` is a tuple that looks like `(n, 1)`, since that is virtually the same thing.

How does this sound?

justinessert replied:
Alternatively, instead of throwing an error, it could raise a warning and set the column's type to object.

It looks like this issue is causing a test failure from this code:
```
data1 = pd.to_datetime("20100101 01:01").tz_localize("UTC")
    df1 = pd.DataFrame(
        columns=["UTCdatetime", "abc"],
    data=data1,
    index=pd.date_range("20140627", periods=1),
 )
```

@TomAugspurger Do you have any guidance on how I should handle this?

TomAugspurger replied:
Hmm the example you posted is a bit unfortunate...  And the fact that it's called this late into DataFrame's constructor makes this really hard. That's already complicated enough.

I would suggest a bit larger refactor if it's OK. I really don't like the idea of `cast_scalar_to_array(shape, ...)` returning anything that *doesn't* have a shape of `shape`.

So, we might have a `cast_scalar_to_extension_arrays(n_rows, scalar, dtype=None)`. And then places that currently use `cast_scalar_to_array` will check the type of the scalar. If it's an extension scalar then we create a list or dict of extension arrays by calling `cast_scalar_to_extension_arrays`.

We'll need to update that test. I'd consider it a bug that the dtype of `Series(data1, index=...)` doesn't match the dtypes of `DataFrame(data1, index=..., columns=...)`.

LMK if any of that is unclear.

justinessert replied:
Thanks @TomAugspurger for your help, that does make sense. I updated the code per these suggestions and everything seems to be working as expected, except the `test_combine_first_timezone` test.

I've gotten pretty lost in debugging this one, I'm not sure why this test should have changed, can you provide some insight here?

TomAugspurger replied:
Mmm this is messy, sorry. But the short version is that the old test was relying on the buggy behavior of `DataFrame(timestamp_scalar)` returning object dtype. Notice the dtypes in the tests

```python
(Pdb) df1.dtypes
UTCdatetime    object
abc            object
dtype: object
(Pdb) df2.dtypes
UTCdatetime    object
xyz            object
dtype: object
```

On this PR those are (correctly) 

```python
(Pdb) df1.dtypes
UTCdatetime    datetime64[ns, UTC]
abc            datetime64[ns, UTC]
dtype: object
(Pdb) df2.dtypes
UTCdatetime    datetime64[ns, UTC]
xyz            datetime64[ns, UTC]
dtype: object
```

Now in `combine` you get into some messy dtype handling issues. In order to not get hung up on those, I'd recommend that you

1. update the tests in `test_combine_first_timezone` to `.astype(object)` to match the old behavior
2. Open an issue demonstrating the buggy behavior (search to make sure we don't have one yet).
3. Add a comment referencing that issue.

justinessert replied:
Thanks for the info, that worked!

TomAugspurger commented on pandas/tests/dtypes/cast/test_infer_dtype.py:
```diff
@@ -187,14 +188,29 @@ def test_infer_dtype_from_array(arr, expected, pandas_dtype):
+    "obj,dtype",
+    [
+        (Period("2011-01-01", freq="D"), PeriodDtype('D')),
+        (Period("2011-01", freq="M"), PeriodDtype('M')),
```
I don't think we need multiple tests for Periods here. It would be useful to have tests for other extension dtypes.

1. `Interval`.
2. `Timestamp` (with and without a timezone)

TomAugspurger replied:
Can you also add "user-facing" tests in `pandas/tests/frame/indexing/test_setitem.py` that do something like the example from the issue?

justinessert replied:
I added an Interval value to the existing test and added a user-facing test in `test_setitem`.

I actually avoided making my `test_cast_scalar_to_extension_array` include Timestamp variables since those are actually covered currently in the `test_cast_scalar_to_numpy_array` test that is above mine. I think the logic here is that timestamps can be converted to numpy timestamp and thus don't need to use an extension array. Please let me know if you would prefer this another way.

TomAugspurger replied:
> timestamps can be converted to numpy timestamp and thus don't need to use an extension array

timezone-naive timestamps can be converted to numpy. But Timestamps with a timezone can't be, and so we have an extension array for tz-aware timestamps.

justinessert replied:
Got it, added that to a Timestamp with a timezone to the test!

TomAugspurger commented on pandas/tests/dtypes/cast/test_infer_dtype.py:
```diff
@@ -187,14 +188,29 @@ def test_infer_dtype_from_array(arr, expected, pandas_dtype):
+    ],
+)
+def test_cast_scalar_to_extension_array(obj, dtype):
+    shape = 3
```
Add a comment like `# GH: <issue number>`

justinessert replied:
Done

TomAugspurger reviewed (commented):
Thanks. Can you add a release note to 1.1.0.rst under bug fixes?

justinessert committed:
Apply suggestions from code review

Checking if extension dtype via built in function instead of manually

Co-authored-by: Tom Augspurger <TomAugspurger@users.noreply.github.com>

justinessert committed:
added missing :

justinessert committed:
modified cast_extension_scalar_to_array test to include an Interval type

justinessert committed:
added user-facing test for extension type bug

justinessert committed:
fixed pep8 issues

justinessert committed:
added note about bug in setting series to scalar extension type

justinessert committed:
corrected order of imports

justinessert committed:
corrected order of imports

justinessert committed:
fixed black formatting errors

justinessert committed:
removed extra comma

justinessert committed:
updated cast_scalar_to_arr to support tuple shape for extension dtype

justinessert committed:
removed unneeded code

justinessert committed:
added coverage for datetime with timezone in extension_array test

justinessert committed:
added TODO

justinessert committed:
correct line that was too long

justinessert committed:
fixed dtype issue with tz test

justinessert commented:
Thanks @TomAugspurger for all your time in reviewing this PR and the original issue. Everything looks to be passing and I think I have addressed all your comments, please lmk if you would like to see any other changes!

jreback commented on pandas/core/dtypes/cast.py:
```diff
@@ -1505,12 +1505,20 @@ def cast_scalar_to_array(shape, value, dtype: Optional[DtypeObj] = None) -> np.n
+    if is_extension_array_dtype(dtype):
+        if isinstance(shape, int):
+            shape = (shape, 1)
+        value = [construct_1d_arraylike_from_scalar(value, shape[0], dtype)]
```
no please
don’t do this here

construct_1s_arraylike can be updated to handle this if needed

TomAugspurger replied:
What's the requested change? We don't want to change `construct_1d_arraylike_from_scalar` to return a list of arrays.

justinessert replied:
@jreback Happy to make whatever changes are needed, but also want to clarify your request.

Currently `construct_1d_arraylike_from_scalar` takes in value, length, and dtype. It then returns a "1d" array similar to `[value] * length`.

Are you suggesting that we change `construct_1d_arraylike_from_scalar` to something like `construct_arraylike_from_scalar` (or maybe `construct_matrixlike_from_scalar`) and allow it to accept a shape (tuple) argument in addition to a length (int) argument?

TomAugspurger replied:
Hopefully https://github.com/pandas-dev/pandas/pull/34875#discussion_r446180430 helps here. Ideally we have

1. One function for casting a scalar to a 1D extension array, which we call several times to create a list of extension ararys
2. One function for casting a scalar to a 2D ndarray (we already have that).

jreback commented on pandas/core/frame.py:
```diff
@@ -528,9 +529,15 @@ def __init__(
+                else:
+                    # Case 2: values is a numpy array
+                    mgr = init_ndarray(
+                        values, index, columns, dtype=values.dtype, copy=False
```
please do not do this here

you can update init_array if needed

TomAugspurger replied:
I don't think we want this change in `init_ndarary`. Right now it turns a list of extension arrays into an ndarray, which we don't want here. That means we'd have the re-scan all the arrays in `values` to ensure that they're all EAs before calling `arrays_to_mgr`.

Where in a unique case here where we know that we have a list of EAs all with the same shape and type.

justinessert replied:
Yeah it seems to me like changing `init_ndarray` would be require more complex handling, but I'm willing to make the change if that's what's requested. If so, I think I would benefit from a more concrete description of what you want, I'm not completely seeing it right now. Thanks!

jreback replied:
Please move this whole section to a separate function, This is make an already really long function REALLY long.

TomAugspurger replied:
This will need to stay here (https://github.com/pandas-dev/pandas/pull/34875#discussion_r446180430). We're going to have to send EA scalars down `ararys_to_mgr` and "ndarray" scalars down `init_ndarary`. I think that's unavoidable.

jreback reviewed (changes requested):
only a brief look
but this patch needs some work

too much special casing

TomAugspurger commented on pandas/core/dtypes/cast.py:
```diff
@@ -1505,12 +1505,20 @@ def cast_scalar_to_array(shape, value, dtype: Optional[DtypeObj] = None) -> np.n
+        if isinstance(shape, int):
+            shape = (shape, 1)
+        value = [construct_1d_arraylike_from_scalar(value, shape[0], dtype)]
+        values = value * shape[1]
```
@justinessert this results in a dataframe where all the columns are the same object. So that something like

```python
df = pd.DataFrame(pd.Timetsamp('2000', tz='CET'), index=[0, 1], columns=[0, 1])
df.iloc[0, 0] = pd.Timestamp("2001")
```

would mutate every column. Try

```python
value = [construct_1d_arraylike_from_scalar(value, shape[0], dtype) for _ in range(shape)]
```

justinessert replied:
Sorry @TomAugspurger I don't understand, this works as I would expect it to. Can you rephrase the issue?

My testing:

```python
df = pd.DataFrame(pd.Timestamp('2000', tz='CET'), index=[0, 1], columns=[0, 1])
```

^Works as expected, both columns 0 and 1 are `datetime64[ns, CET]`

```python
df1 = df.copy()
df1.iloc[0,0] = pd.Timestamp("2001")
```

^This one shouldn't even call the cast scalar to array function I believe, since it is only assigning a single value, not a full column. But this also works as expected for me, column 1 remains `datetime64[ns, CET]` and column 0 becomes `object` (since it has mixed types).

Other similar cases:

```python
df2 = df.copy()
df2[0] = pd.Timestamp("2001")
```

^This does call cast_scalar_to_array and also works as expected I believe. Column 0 becomes `datetime64[ns]` and column 1 remains `datetime64[ns, CET]`

```python
df3 = df.copy()
df3.iloc[0] = pd.Timestamp("2001")
```

^This does call cast_scalar_to_array and also works as expected I believe. Both columns become `object` types (because they have mixed types)

TomAugspurger replied:
Sorry, bad example. The value being set should have had the same timezone. Does this help?

```python
In [13]: df = pd.DataFrame(pd.Timestamp('2000', tz='CET'), index=[0, 1], columns=[0, 1])
    ...:

In [14]: df[0].array is df[1].array
Out[14]: True

In [16]: df[0].array[0] = pd.Timestamp("2001", tz="CET")

In [17]: df
Out[17]:
                          0                         1
0 2001-01-01 00:00:00+01:00 2001-01-01 00:00:00+01:00
1 2000-01-01 00:00:00+01:00 2000-01-01 00:00:00+01:00
```

Notice that the `1` column was mutated too. We want to create a distinct array for each column, not a single array each column refers to.

justinessert replied:
Ahh, yes that makes complete sense. Updated

justinessert committed:
creating distinct arrays for each column

justinessert committed:
resolving mypy error

jorisvandenbossche commented on pandas/core/dtypes/cast.py:
```diff
@@ -1505,10 +1505,20 @@ def cast_scalar_to_array(shape, value, dtype: Optional[DtypeObj] = None) -> np.n
+        if isinstance(shape, int):
+            shape = (shape, 1)
+        return [
+            construct_1d_arraylike_from_scalar(value, shape[0], dtype)
```
Can you update the docstring about this new return type?

jorisvandenbossche replied:
Also, why is it actually needed to return a list if the specified shape is 1D?

justinessert replied:
For some reason I thought that wouldn't work, but it looks like it does. Thanks for the suggestion
Also added info in the docstring
Also added constructor test

justinessert replied:
It's a list of 1d extension arrays, ie [PeriodArray, PeriodArray] (I updated the docstring to reflect this)

jorisvandenbossche commented on pandas/core/frame.py:
```diff
@@ -528,9 +529,15 @@ def __init__(
+                if isinstance(values, list):
+                    # Case 1: values is a list of extension arrays
+                    dtype, _ = infer_dtype_from_array(values[0], pandas_dtype=True)
+                    mgr = arrays_to_mgr(values, columns, index, columns, dtype=dtype)
```
Why do you need to get the dtype and pass it to `arrays_to_mgr` here? You already know you have a list of ExtensionArrays (which are properly typed), right? So if you don't pass a dtype, `arrays_to_mgr` should just take the dtype of the passed arrays I think?

justinessert replied:
For some reason I thought that wouldn't work, but it looks like it does. Thanks for the suggestion

jorisvandenbossche reviewed (commented):
Thanks for working on this!

Is the change in the DataFrame constructor also tested?

justinessert committed:
added docstring info and test

justinessert committed:
removed unneeded import

justinessert commented:
@TomAugspurger @jorisvandenbossche @jreback Are there any outstanding requests that you have on this PR?

jreback commented:
i will have a look soon as havent  reviewed

jreback commented on pandas/core/dtypes/cast.py:
```diff
@@ -1494,21 +1494,32 @@ def cast_scalar_to_array(shape, value, dtype: Optional[DtypeObj] = None) -> np.n
     Returns
     -------
-    ndarray of shape, filled with value, of specified / inferred dtype
+    ndarray of shape of list of length shape[1] of Extension Arrays of length shape[0],
```
really? why can't we just have an int and the caller can pass shape[1] if needed. this is *really* *really* confusing.

justinessert replied:
To be clear, this is not something that I am changing via this PR, this function previously accepted a `shape` argument, not a `length` argument.

I can't be sure of the motivation here, but I'd guess that this is related to how pandas manages the data in memory. If you have 5 int columns in a DataFrame, you don't want those to be in 5 distinct series since that creates a lot of overhead. Instead you create a single Numpy array of width 5.

jreback commented on pandas/core/dtypes/cast.py:
```diff
@@ -1494,21 +1494,32 @@ def cast_scalar_to_array(shape, value, dtype: Optional[DtypeObj] = None) -> np.n
     else:
         fill_value = value
 
+    # TODO: Update this function to add support for 3rd party extension types
```
why is this a TODO? isn't this exactly what this PR is doing?

justinessert replied:
Not actually too sure on this one, @TomAugspurger pointed out that the code may not support 3rd party extension types.

But no, that is not the purpose of this PR. The purpose of this PR is to support 1st part extension types (i.e. Period, Timestamp, Interval, etc)

TomAugspurger replied:
The linked issue has more information. But the summary is that `infer_dtype_from_scalar` doesn't work for arbitrary scalars that might be contained in a 3rd party extension type.

jreback commented on pandas/core/frame.py:
```diff
@@ -528,9 +528,14 @@ def __init__(
-                mgr = init_ndarray(
-                    values, index, columns, dtype=values.dtype, copy=False
-                )
+                if isinstance(values, list):
```
this is a crazy api here. Why are you doing this?

justinessert replied:
I try to clarify this in the comments, but I can understand why this could be confusing. There are two distinct possibilities:

1. `data` is a non-extension type, in which case `cast_scalar_to_array` will return a numpy ndarray
2. `data` is an extension type, in which cast `cast_scalar_to_array` will return a list of extension arrays

The two cases also require two separate instantiations of the manager, since they are different data types.

I'm not sure we can do better here. If `data` is an extension type, then it can't be put into a Numpy array and if `data` is not an extension type, we don't want to create a list of separate arrays, we want to create a single higher-dimensional array

TomAugspurger replied:
Ideally the flow is something like

```python
# in DataFrame.__init__
if values_is_a_scalar:  # the case we're in here
    # discover if we're an extension dtype (make a list of arrays)
    # or a numpy dtype (make a single ndarray)
    dtype = dtype or infer_dtype(values, pandas_dtype=True)  # infer if needed
    if is_extension_array_dtype(dtype):
        values = cast_scalar_to_list_of_extension_arrays(...)
        mgr = arrays_to_mgr(...)
    else:
        values = cast_scalar_to_array(...)
        mgr = init_ndarray(...)
```


So the only extra complexity in `DataFrame.__init__` reflects the *fundamental* fact that an "ndarray" scalar has to take a different path from the "extension" scalar.

However, I think that right now `dtype` might be `None`, since `cast_array_to_scalar` does a bit of inference. So ideally `cast_array_to_scalar` only does the array construction, and doesn't do the dtype inference. Does that make sense @justinessert?

justinessert replied:
@TomAugspurger Happy to follow all or some of that method. I also just pushed a way to flatten out the `else` case, curious if you want to keep any element of this?
```
# DOESN'T WORK FOR TIMEZONE, PLEASE SUGGEST BETTER WAY TO DO THIS CHECK
        elif isinstance(data, (Period, Interval)):
            values = cast_scalar_to_array((len(index), len(columns)), data, dtype=dtype)

            if index is not None and columns is not None:
                mgr = arrays_to_mgr(values, columns, index, columns, dtype=None)
            else:
                raise ValueError("DataFrame constructor not properly called!")
        else: # Else case Identical to the way it was previously!
            try: ...
            except (ValueError, TypeError) as err: ...

            if arr.ndim == 0 and index is not None and columns is not None:
                values = cast_scalar_to_array(
                    (len(index), len(columns)), data, dtype=dtype
                )
                mgr = init_ndarray(
                    values, index, columns, dtype=values.dtype, copy=False
                )
            else:
                raise ValueError("DataFrame constructor not properly called!")
```

TomAugspurger replied:
Hmm hard to say, but I don't think we want to be treating Period / Interval specially. Ideally we're just looking at the dtype implied by the scalar and using that to determine which path we take.

jreback commented on pandas/core/frame.py:
```diff
@@ -3731,6 +3736,11 @@ def reindexer(value):
             # upcast
             value = cast_scalar_to_array(len(self.index), value)
+
+            # if extension dtype, value will be a list of length 1
```
what is this?

justinessert replied:
In this case, we simply want a 1d array, so again our two cases are:

1. `data` is a non-extension type, in which case `cast_scalar_array` will return a numpy array of width 1, which is what we want, in this case we don't need to do anything
2. `data` is an extension type, in which case `cast_scalar_array` will return a list containing a single extension array. In this case, we just want the extension array, so we index into the list and grab the only element.

jreback reviewed (changes requested):
this is a bandaid and needs a serious refactor. sorry to be blunt, but this is adding a bunch of if/thens which likey are buggy and special cases. This needs a real solution and not more hacks.

justinessert committed:
flattened else case in init

justinessert committed:
refactored extension type column fix

justinessert committed:
reverted docstring changes

justinessert committed:
reverted docstring changes

justinessert committed:
removed unneeded imports

justinessert committed:
reverted test changes

justinessert commented:
@TomAugspurger @jreback I refactored this PR per your suggestions. In affect I am no longer changing the functionality of `cast_scalar_to_array` and everything is a lot more explicit. Please let me know what you think now.

justinessert committed:
fixed construct_1d_arraylike bug

jreback commented on pandas/core/frame.py:
```diff
@@ -515,24 +516,38 @@ def __init__(
-                    values, index, columns, dtype=values.dtype, copy=False
-                )
+            if is_extension_array_dtype(dtype):
+                if index is None or columns is None:
```
can you move this check higher before this if ?

jreback commented on pandas/core/frame.py:
```diff
@@ -515,24 +516,38 @@ def __init__(
-                mgr = init_ndarray(
-                    values, index, columns, dtype=values.dtype, copy=False
-                )
+            if is_extension_array_dtype(dtype):
```
can you give some comments on wha tis happening here

jreback commented on pandas/core/frame.py:
```diff
@@ -515,24 +516,38 @@ def __init__(
+                mgr = arrays_to_mgr(values, columns, index, columns, dtype=None)
             else:
-                raise ValueError("DataFrame constructor not properly called!")
+                try:
```
can you comment on what is happening here

also this is basically a straight move?

justinessert replied:
Yeah, this already existed here, not totally sure what the intent is. If I had to guess, I'd say it's creating an array and then a few lines down checks to see if the arrays dimensionality is 0, in effect confirming that data is a scalar

jreback replied:
if this is a straight move that is ok. just a general comment, like 'attempt to coerce to a numpy array'

justinessert replied:
@jreback I see this response, but do you have anything else that you want to see in this PR? Would like to group some other changes together, if necessary, so I don't have to create a commit just to add a single comment haha

justinessert replied:
Added this comment

jreback commented on pandas/core/frame.py:
```diff
@@ -3730,7 +3745,13 @@ def reindexer(value):
+                    value, len(self.index), infer_dtype
+                )
+            else:
+                value = cast_scalar_to_array(len(self.index), value)
```
can you not do this *inside* cast_scalar_to_array (pass in the `infer_dtype` optionally)

justinessert replied:
Can you clarify what you mean by "this"? Your last comments were to move this whole if/else logic out of `cast_scalar_to_array`, which I do agree made for a much cleaner solution

jreback replied:
instead of adding this if/then clause, can you change `cast_scalar_to_array` to just handle if the dtype is an extension_array? (and obviously pass infer_dtype here)

jorisvandenbossche replied:
`cast_scalar_to_array` only deals with numpy arrays. Initially this PR added handling extension dtypes in there as well, but after discussion we actually moved it out because it gave a lot of complications (eg EAs don't support 2D arrays, while `cast_scalar_to_array` generally does).

jreback replied:
exactly the reason to make the change. we are almost certainly going to need to do this in other places and this is a natural fit.

jorisvandenbossche replied:
@jreback then please first answer your own comment at https://github.com/pandas-dev/pandas/pull/34875#discussion_r445880416 on how to solve it differently (and how to deal with 2D shape in `cast_scalar_to_array` in case of extension dtypes). Moving it out of `cast_scalar_to_array` was exactly to avoid this "crazy api" you commented on.

justinessert replied:
I agree that that would be ideal, but I believe that would be a much larger refactor and to me it seems like something that should have it's own PR.

I do acknowledge that this PR is a band-aid that may not be required if we were to make that change. But I still think it would be best to apply the band-aid and come back to that at a later time given the maturity of this PR and the complexity associated with that refactor.

Though if this is something that we want to put in this PR, I would benefit from some fairly verbose assistance in coding that up. I'm not sure that I have the context and code familiarity to do that alone.

justinessert replied:
@jreback What are your thoughts here?

jreback replied:
how does this involve 2D at all here? simply move the if_extenseion array case *into* `cast_scalar_arry`. The constructor here is WAY to big to grok, adding if/else cases makes this much worse.

jorisvandenbossche replied:
``cast_scalar_array`` in general involves 2D (not here, but it is also used a bit higher in this file in the DataFrame constructor with a 2D shape, and a similar if/else structure is used there). If we let it handle extension arrays, we should let it handle those for all shapes IMO, otherwise it is making that only more confusing.

Again, please look at your own earlier comment (https://github.com/pandas-dev/pandas/pull/34875#discussion_r445880416 and the answers to that) and propose a way how to deal with this in `cast_scalar_array` without a "crazy api" (which was returning a list of 1D EAs for the 2D case).

jorisvandenbossche replied:
If you want to see the state at the point when things were handled inside `cast_scalar_array`, this is the diff of the commits before it was refactored: https://github.com/pandas-dev/pandas/pull/34875/files/90a85706ce1bfa1735cf07190dc61efaa62e454d

jreback replied:
I see the code and this is easily possible. please try to do as I have described. I don't see any reason not to.

justinessert replied:
@jreback Please be more specific. What I would do is exactly as I had it, which you called a "crazy api" and specifically requested I change.
Are you asking me to revert the changes that you requested? Or are you after something else?

justinessert replied:
To be more specific, it would be easy if `cast_scalar_array` was simply returning a single 1d array. But in most cases it is not, it is returning a 2d array.
Since there's no good way to make a 2d Extension array, I handled this by returning a list of Extension arrays. Is this what your asking I go back to doing?

justinessert replied:
We cannot "simply move the if_extenseion array case into cast_scalar_arry", because this is a specific case where `cast_scalar_array` is receiving an int to its `shape` parameter. But in general, `shape` is a tuple to define the shape of the 2d array.

jreback replied:
https://github.com/pandas-dev/pandas/blob/master/pandas/core/dtypes/cast.py#L1492

you just need to move the if_extension_dtype *into* this function. I don't think there is ambiguity or anything. This returns exactly 1 ndarray. I am not sure where 2D even matters here at all.

justinessert replied:
A single ndarray can be 2 dimensional, and in this case it often is, for example, when it is used [here].(https://github.com/pandas-dev/pandas/blob/c73a2dee55f48690ad9a442ba019f9bf3bc22c26/pandas/core/frame.py#L552)

Doing this in `cast_scalar_to_array` would NOT work without a larger refactor.
```
if is_extension_array_dtype(dtype):
    value = construct_1d_arraylike_from_scalar(fill_value, shape, dtype)
else:
    values = np.empty(shape, dtype=dtype)
    values.fill(fill_value)
```
this `shape` parameter is usually a **tuple** not an **int** like it is here, so `construct_1d_arraylike_from_scalar` would fail when `shape=(2, 5)` for example (since extension arrays are only 1d)

justinessert replied:
"This returns exactly 1 ndarray" is technically true, but it can be a 2d ndarray

For example
```python
values = np.empty((2,3), dtype=int)
values.fill(2)
```
would give you a numpy array like
```
[[2, 2, 2], [2, 2, 2]]
```

jreback commented on pandas/tests/frame/methods/test_combine_first.py:
```diff
@@ -217,10 +219,13 @@ def test_combine_first_timezone(self):
-        tm.assert_frame_equal(res, exp)
         assert res["UTCdatetime"].dtype == "datetime64[ns, UTC]"
         assert res["abc"].dtype == "datetime64[ns, UTC]"
+        # GH Issue 7509
```
can you split this to a separate test (the added code and what is needed to repro)

jorisvandenbossche replied:
Is not possible to remove all added `dtype="object"` ? The `df1` and `df2` now changed to no longer have object dtype due to the fix, but this actually seems a more correct case to test. Or does `combine_first` not handle that correctly / still return object?

TomAugspurger replied:
Fixing the setitem bug made these datetime64ns rather than object. But then `combine_first` is buggy, as described in https://github.com/pandas-dev/pandas/issues/7509.

So yes, ideally we test non-object, but https://github.com/pandas-dev/pandas/issues/7509 will have to be fixed first (and shouldn't block this PR).

jorisvandenbossche replied:
Ah, thanks for the pointer. Yes certainly shouldn't block the PR. 
It might be worth adding a comment explaining this here though (instead of just the github issue reference, as you need to look that up to know what it is about)

justinessert replied:
Thanks for the recommendation, updated the comment

jreback commented on pandas/tests/frame/test_constructors.py:
```diff
@@ -723,6 +724,12 @@ def test_constructor_period(self):
         assert df["a"].dtype == a.dtype
         assert df["b"].dtype == b.dtype
 
+        data = pd.Period("2012-01", freq="M")
```
can you add the issue number and make a separate test. Also use a full comparison, e.g. assert_frame_equal here

TomAugspurger replied:
Writing out the full expected and using `tm.assert_frame_equal` will fit better with our other tests, and helps prevent regressions in future refactors.

justinessert replied:
Gotcha, done!

jreback replied:
can you parameterize on all scalar types (as you did in the setitem tests).

jreback reviewed (changes requested):
thanks @justinessert looking a lot nicer. This is much more in line with what was looking for. a few additional comments.

jreback commented on doc/source/whatsnew/v1.0.0.rst:
```diff
@@ -1259,6 +1259,7 @@ ExtensionArray
 - Bug in :class:`arrays.PandasArray` when setting a scalar string (:issue:`28118`, :issue:`28150`).
 - Bug where nullable integers could not be compared to strings (:issue:`28930`)
 - Bug where :class:`DataFrame` constructor raised ``ValueError`` with list-like data and ``dtype`` specified (:issue:`30280`)
+- Bug where :class:`Series` set to scalar extension type was considered an object type rather than the extension type (:issue:`34832`)
```
since we had to change some other tests, I think we need to break this to a new section and show the changes from before. E.g. construction of a multi-column df now is object if we don't have unform datetimes? (we need to be very clear what is the change here since we had to change some tests)

justinessert replied:
Realized that I added this to the wrong file, so I'm moving this addition to `v1.1.0.rst`. But can you clarify what you would like me to do here? I'm not totally sure based on your comment.

I think that this line correctly describes the change. Are you asking to also include an example, such as

```
# This used to create an object type column, now it creates a Period type column
DataFrame(index=[0, 1], columns=["a"], data=pd.Period("2020-01"))
```

The example you gave, where a datetime column has multiple different timezones, this was always an object column.

jreback replied:
ok tell you what. let's create an issue to refactor this (frame constructor). That's what I mainly have a problem with, we are adding if/then ALL over the place for extension types rather than a proper refactor.

So ok with merging this (just one small change on the naming in the tests). And please create an issue (and if you want / can refactor would be great) as a followup.

justinessert committed:
reorganized if statements

justinessert committed:
moved what's new statement to correct file

justinessert committed:
created new test for period df construction

TomAugspurger reviewed (approved):
Just one comment on the test, but I think the change here look good.

justinessert committed:
added assert_frame_equal to period_data test

justinessert commented:
Can someone help me understand what this recent test failure is? I could be wrong, but it doesn't seem to be a unit test or formatting error

TomAugspurger commented:
I restarted the job that crashed.

jorisvandenbossche commented on pandas/tests/frame/test_constructors.py:
```diff
@@ -723,6 +724,18 @@ def test_constructor_period(self):
+        assert df["a"].dtype == PeriodDtype("M")
+        assert df["b"].dtype == PeriodDtype("M")
+
+        expected = DataFrame({"a": [data] * 2, "b": [data] * 2})
```
```suggestion
        arr = pd.array([data] * 2, dtype=PeriodDtype("M")
        expected = DataFrame({"a": arr, "b": arr})
```

(this should give a bit more confidence in the test IMO, as then the construction of the expected result doesn't rely on inference of (a list of) scalars as well)

jorisvandenbossche commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -1078,6 +1078,7 @@ ExtensionArray
 - Fixed bug where :meth:`StringArray.memory_usage` was not implemented (:issue:`33963`)
 - Fixed bug where :meth:`DataFrameGroupBy` would ignore the ``min_count`` argument for aggregations on nullable boolean dtypes (:issue:`34051`)
 - Fixed bug that `DataFrame(columns=.., dtype='string')` would fail (:issue:`27953`, :issue:`33623`)
+- Bug where :class:`Series` set to scalar extension type was considered an object type rather than the extension type (:issue:`34832`)
```
I find this sentence a bit hard to understand ("Series set to scalar extension type" -> it is not a Series directly that we are setting, but actually a column in a dataframe). Can you try to rephrase?

jorisvandenbossche replied:
For creating a Series directly this actually also fails on master:

```
In [14]: pd.Series(pd.Period("2012", freq="A"), index=[0, 1])   
Out[14]: 
0    2012
1    2012
dtype: object
```

and I assume this is not fixed by this PR (but fixing that is for a separate PR! ;-))

justinessert replied:
Good point! Updated the wording

jorisvandenbossche reviewed (approved):
A few small comments, but generally looks good to me now!

justinessert committed:
Using pandas array instead of df constructor for better test

Co-authored-by: Joris Van den Bossche <jorisvandenbossche@gmail.com>

justinessert committed:
changed wording

justinessert committed:
Merge branch 'master' of https://github.com/justinessert/pandas

justinessert committed:
pylint fixes

jorisvandenbossche commented:
@jreback all good?

jreback commented on pandas/tests/frame/indexing/test_setitem.py:
```diff
@@ -150,3 +161,22 @@ def test_setitem_dict_preserves_dtypes(self):
             }
         tm.assert_frame_equal(df, expected)
+
+    def test_setitem_extension_types(self):
```
can you parameterize on scalars here (period, interval) also add Timestamp(tz=....)

justinessert replied:
Done

jreback reviewed (changes requested):
sorry thought i left these comments a while back

justinessert committed:
parameterized test and added comment

justinessert committed:
removed extra comma

justinessert committed:
Merge branch 'master' into master

justinessert committed:
parameterized test

jreback commented on pandas/tests/frame/test_constructors.py:
```diff
@@ -713,6 +716,29 @@ def test_constructor_period(self):
+            ),
+        ],
+    )
+    def test_constructor_period_data(self, data, dtype):
```
can you change the name -> test_constructor_scalar_data

jreback replied:
@justinessert here.

justinessert replied:
done

jreback commented:
ping on green.

justinessert committed:
renamed test

jreback commented:
thanks @justinessert

jorisvandenbossche commented:
Thanks a lot @justinessert ! (and sorry for the difficult back and forth discussion)

justinessert commented:
No problem!

@jorisvandenbossche @jreback @TomAugspurger Thanks for all your help in getting this PR through!
